include/logging/log_msg.h

```diff
@@ -0,0 +1,617 @@
+	      "Structure must fit in 2 bytes");
+
+/** Part of log message header specific to hexdump log message. */
+struct log_msg_hexdump_hdr {
+	COMMON_PARAM_HDR();
+	u16_t raw_string : 1;
+	u16_t length     : LOG_MSG_HEXDUMP_LENGTH_BITS;
+};
+
+static_assert(sizeof(struct log_msg_hexdump_hdr) == sizeof(u16_t),
+	      "Structure must fit in 2 bytes");
+
+/** Log message header structure */
+struct log_msg_hdr {
+	atomic_t ref_cnt;       /*!< Reference counter used for tracking number of message users. */
+	union {
+		struct log_msg_generic_hdr generic;
+		struct log_msg_std_hdr std;
+		struct log_msg_hexdump_hdr hexdump;
+		u16_t raw;
+	} params;
+	struct log_msg_ids ids; /*!< Identification part of the message.*/
+	u32_t timestamp;        /*!< Timestamp. */
+};
+
+/** @brief Data part of head of standard log message. */
+struct log_msg_std_head_data {
+	const char *str;
+	u32_t args[LOG_MSG_NARGS_SINGLE_CHUNK];
+};
+
+/** @brief Data part of head of extended standard log message.
+ *
+ * @details When message is extended, head of the message contains also a
+ *          pointer to the next chunk thus data part is reduced.
+ */
+struct log_msg_std_ext_head_data {
+	const char *str;
+	u32_t args[LOG_MSG_NARGS_HEAD_CHUNK];
+};
+
+/** @brief Data part of log message. */
+union log_msg_head_data {
+	struct log_msg_std_head_data std;
+	u8_t bytes[LOG_MSG_HEXDUMP_BYTES_SINGLE_CHUNK];
+};
+
+/** @brief Data part of extended log message. */
+struct log_msg_ext_head_data {
+	struct log_msg_cont *next;
+	union {
+		struct log_msg_std_ext_head_data std;
+		u8_t bytes[LOG_MSG_HEXDUMP_BYTES_HEAD_CHUNK];
+	} data;
+};
+
+/** @brief Log message structure. */
+struct log_msg {
+	struct log_msg *next;   /*!< Used by logger core to put message into the list.*/
+	struct log_msg_hdr hdr; /*!< Message header. */
+
+	union {
+		union log_msg_head_data data;
+		struct log_msg_ext_head_data ext;
+	} data;                 /*!< Message data. */
+};
+
+static_assert(sizeof(union log_msg_head_data) ==
+	      sizeof(struct log_msg_ext_head_data),
+	      "Structure must be same size");
+
+/** @brief Chunks following message head when message is extended. */
+struct log_msg_cont {
+	struct log_msg_cont *next;                                      /*!< Pointer to the next chunk. */
+	union {
+		u32_t args[LOG_MAX_NARGS - LOG_MSG_NARGS_HEAD_CHUNK];   /*!< Standard message extension. */
+		u8_t bytes[HEXDUMP_BYTES_CONT_MSG];                     /*!< Hexdump message extension. */
+	} data;
+};
+
+
+
+/** @brief Log message */
+union log_msg_chunk {
+	struct log_msg head;
+	struct log_msg_cont cont;
+};
+
+extern struct k_mem_slab log_msg_pool;
+
+/** @brief Function for indicating that message is in use.
+ *
+ *  @details Message can be used (read) by multiple users. Internal reference
+ *           counter is atomically increased. See @ref log_msg_put.
+ *
+ *  @param msg Message.
+ */
+void log_msg_get(struct  log_msg *msg);
```

Comment:
Double-space appear here and in other places - not in Zephyr codestyle.

Reply:
sure, fixed. Unfortunately, uncrustify did not handled that.